src/oss-download/PackageDownloader.cs

```diff
@@ -0,0 +1,258 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CST.OpenSource.Shared;
+
+namespace Microsoft.CST.OpenSource
+{
+    /// <summary>
+    /// Class for managing the download of a single package
+    /// </summary>
+    public class PackageDownloader
+    {
+
+        /// <summary>
+        /// Logger for each of the subclasses
+        /// </summary>
+        protected static readonly NLog.Logger Logger = NLog.LogManager.GetCurrentClassLogger();
+
+        List<PackageURL> PackageVersions { get; set; }
+
+        BaseProjectManager packageManager { get; set; }
+
+        string destinationDirectory { get; set; }
+
+        // should we cache/check for the cache?
+        bool doCache { get; set; }
+
+        // folders created
+        List<string> downloadPaths { get; set; }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="args">parameters passed in from the user</param>
+        public PackageDownloader(PackageURL purl, string destinationDir)
+        {
+            if (purl == null)
+            {
+                throw new ArgumentNullException("purl cannot be null");
+            }
+
+            this.destinationDirectory = destinationDir;
+            // if we are told to use a cache dir, and it exists, believe that caching is still doable
+            if (string.IsNullOrEmpty(destinationDir))
+            {
+                this.doCache = false;
```

Comment:
I am not a fan of this pattern.  I think you should explicitly tell the package downloader if downloads should use cache or not - it shouldn't be based on the content of the destinationDirectory argument.